### PR TITLE
Fetch Metadata article for web.dev

### LINKED ISSUE #2972: content: Fetch Metadata [2020-05-25 HARD?]
* [Draft](https://docs.google.com/document/d/1yt4qu0VFgDwb9gF6TZfLAVdzTCxR9KykB4t4_lru0Fw/edit)

### PR DESCRIPTION
**Pull request by lweichselbaum**
This article should serve as a reference for developers on how to use Fetch Metadata request headers to protect web applications from common cross-site attacks.

Fixes #2972

### CODE REVIEW

**Comment by kaycebasques** [comment]
This PR was a bit premature but will eventually be relevant. @rowan-m is reviewing the Google Doc draft.

**Comment by rowan-m** [comment]
Left a tiny update on the draft, but aside from that LGTM.

**Comment by lweichselbaum** [comment]
Thanks everyone! I upstreamed the latest changes from the doc. ptal.

File: src/site/content/en/blog/fetch-metadata/index.md [hunk 1]
```diff
@@ -123,32 +135,41 @@ To ensure that your site can still be linked from other sites, you have to allow
...
 
-#### Step 4: (Optional) Opt out endpoints that are meant to serve cross-site traffic
+#### Step 4: Opt out endpoints that are meant to serve cross-site traffic (Optional) 
+
 In some cases, your application might provide resources which are meant to be loaded cross-site. These resources need to be exempted on a per-path or per-endpoint basis. Examples of such endpoints are:
- - CORS-enabled endpoints
- - Public resources like stylesheets or images
+
+ - Endpoints meant to be accessed cross-origin - If your application is serving endpoints that are `CORS` enabled, you need to explicitly opt them out from resource isolation to ensure that cross-site requests to these endpoints are still possible.
+ - Public resources (e.g. images, styles, etc) - Any public and unauthenticated resources that should be loadable cross-origin from other sites can be exempted as well.
```
Comment (jpmedley) [inline src/site/content/en/blog/fetch-metadata/index.md:147]:
```suggestion
 - Public resources (e.g. images, styles, etc.) - Any public and unauthenticated resources that should be loadable cross-origin from other sites can be exempted as well.
```

File: src/site/content/en/blog/fetch-metadata/index.md [hunk 2]
```diff
@@ -123,32 +135,41 @@ To ensure that your site can still be linked from other sites, you have to allow
...
+
+```python
+  if req.path in ('/my_CORS_endpoint', '/favicon.png'):
+    return True
```
Comment (jpmedley) [inline src/site/content/en/blog/fetch-metadata/index.md:151]:
```suggestion
if req.path in ('/my_CORS_endpoint', '/favicon.png'):
  return True
```

Comment (jpmedley) [reply src/site/content/en/blog/fetch-metadata/index.md:151]: Can you please remove the indentation from all code examples? It's causing the page to render with the wrong left-hand spacing.

File: src/site/content/en/blog/fetch-metadata/index.md [hunk 3]
```diff
@@ -123,32 +135,41 @@ To ensure that your site can still be linked from other sites, you have to allow
...
- - **Modifying the request**. In certain scenarios, consider performing other actions like redirecting to your landing page and dropping authentication credentials (e.g. cookies). However, please be aware that this could weaken the protections of a Fetch Metadata-based policy.
+ - **Modifying the request**. In certain scenarios, consider performing other actions like redirecting to your landing page and dropping authentication credentials (e.g. cookies). However, be aware that this could weaken the protections of a Fetch Metadata-based policy.
 {% endAside %}
 
 
 **Example:** The following code demonstrates a complete implementation of a robust Resource Isolation Policy on the server or as a middleware to deny potentially malicious cross-site resource requests, while allowing simple navigational requests:
 
 ```python
-# Reject cross-origin requests to protect from CSRF, XSSI & other bugs
+# Reject cross-origin requests to protect from CSRF, XSSI, & other bugs
```
Comment (jpmedley) [inline src/site/content/en/blog/fetch-metadata/index.md:172]:
```suggestion
# Reject cross-origin requests to protect from CSRF, XSSI, and other bugs
```

File: src/site/content/en/blog/fetch-metadata/index.md [hunk 4]
```diff
@@ -164,30 +185,27 @@ def allow_request(req):
...
-1. Fix potential violations by exempting legitimate cross-origin endpoints (e.g. loaded using CORS).
+
+1. Install a module like the code snippet from above to log and monitor how your site behaves and make sure the restrictions don't affect any legitimate traffic.
+1. Fix potential violations by exempting legitimate cross-origin endpoints.
 1. Enforce the policy by dropping non-compliant requests.
 
 ### Identifying and fixing policy violations
-We recommend testing your policy in a side-effect free way by first enabling it in reporting mode in your server-side code. Alternatively, you can implement this logic in middleware, or in a reverse proxy which logs any violations that your policy might produce when applied to production traffic.
+
+It's recommended to test your policy in a side-effect free way by first enabling it in reporting mode in your server-side code. Alternatively, you can implement this logic in middleware, or in a reverse proxy which logs any violations that your policy might produce when applied to production traffic.
```
Comment (jpmedley) [inline src/site/content/en/blog/fetch-metadata/index.md:203]:
```suggestion
It's recommended that you test your policy in a side-effect free way by first enabling it in reporting mode in your server-side code. Alternatively, you can implement this logic in middleware, or in a reverse proxy which logs any violations that your policy might produce when applied to production traffic.
```

File: src/site/content/en/blog/fetch-metadata/index.md [hunk 5]
```diff
@@ -164,30 +185,27 @@ def allow_request(req):
...
 
-However there are a few legitimate cases that are safe to opt out:
-
-- Endpoints meant to be accessed cross-origin - If your application is serving endpoints that are `CORS` enabled, you need to explicitly opt them out from resource isolation to ensure that cross-site requests to these endpoints are still possible.
-
--  Public resources (e.g. images, styles, etc) - Any public and unauthenticated resources that should be loadable cross-origin from other sites can be exempted as well. 
-
 ### Enforcing a Resource Isolation Policy
-After you've checked that your policy doesn't impact legitimate production traffic, you're ready to start enforcing the restrictions, guaranteeing that other sites will not be able to request your resources, protecting your users from cross-site attacks.
+After you've checked that your policy doesn't impact legitimate production traffic, you're ready to start enforcing the restrictions, guaranteeing that other sites will not be able to request your resources and protecting your users from cross-site attacks.
```
Comment (jpmedley) [inline src/site/content/en/blog/fetch-metadata/index.md:208]:
```suggestion
After you've checked that your policy doesn't impact legitimate production traffic, you're ready to enforce restrictions, guaranteeing that other sites will not be able to request your resources and protecting your users from cross-site attacks.
```

File: src/site/content/en/blog/fetch-metadata/index.md [hunk 6]
```diff
@@ -0,0 +1,220 @@
...
+
+Any requests that do not originate from a cross-origin context (like `evil.example`) will be allowed. In particular, these are requests that:
+
+ -  Originate from your own application (e.g. a same-origin request where `site.example` requests `site.example/foo.json` will always be allowed).
+ - Originate from your subdomains.
+ -  Are explicitly caused by a user's interaction with the user agent (e.g. direct navigation or by clicking a bookmark, etc.).
+
+```python
+  if req['sec-fetch-site'] in ('same-origin', 'same-site', 'none'):
+    return True  # Allow this request
```
Comment (kaycebasques) [inline src/site/content/en/blog/fetch-metadata/index.md:120]:
```suggestion
if req['sec-fetch-site'] in ('same-origin', 'same-site', 'none'):
  return True  # Allow this request
```

File: src/site/content/en/blog/fetch-metadata/index.md [hunk 7]
```diff
@@ -0,0 +1,220 @@
...
+
+#### Step 3: Allow simple top-level navigation and iframing
+
+To ensure that your site can still be linked from other sites, you have to allow simple (`HTTP GET`) top-level navigation.  
+
+```python
+  if req['sec-fetch-mode'] == 'navigate' and req.method == 'GET'
+    # <object> and <embed> send navigation requests, which we disallow.
+    and req['sec-fetch-dest'] not in ('object', 'embed'):
+      return True  # Allow this request
```
Comment (kaycebasques) [inline src/site/content/en/blog/fetch-metadata/index.md:135]:
```suggestion
if req['sec-fetch-mode'] == 'navigate' and req.method == 'GET'
  # <object> and <embed> send navigation requests, which we disallow.
  and req['sec-fetch-dest'] not in ('object', 'embed'):
    return True  # Allow this request
```

File: src/site/content/en/blog/fetch-metadata/index.md [hunk 8]
```diff
@@ -0,0 +1,220 @@
...
+
+Since not all browsers support Fetch Metadata, you need to allow requests that don't set `Sec-Fetch-*` headers by checking for the presence of `sec-fetch-site`.
+
+{% Aside %}
+  All of the following examples are Python code.
+{% endAside %}
+
+```python
+  if not req['sec-fetch-site']:
+    return True  # Allow this request
```
Comment (kaycebasques) [inline src/site/content/en/blog/fetch-metadata/index.md:102]:
```suggestion
if not req['sec-fetch-site']:
  return True  # Allow this request
```

File: src/site/content/en/blog/fetch-metadata/index.md [hunk 9]
```diff
@@ -0,0 +1,220 @@
...
+{% Aside 'caution' %}
+Make sure that you reject invalid requests before running authentication checks or any other processing of the request to prevent revealing sensitive timing information. 
+{% endAside %}
+
+
+## Further reading
+
+- [W3C Fetch Metadata Request Headers specification](https://www.w3.org/TR/fetch-metadata/)
+- [Fetch Metadata Playground](https://secmetadata.appspot.com/)
+- [Google I/O - Securing Web Apps with Modern Platform Features](https://webappsec.dev/assets/pub/Google_IO-Securing_Web_Apps_with_Modern_Platform_Features.pdf)
```
Comment (kaycebasques) [inline src/site/content/en/blog/fetch-metadata/index.md:219]:
```suggestion
- [Google I/O - Securing Web Apps with Modern Platform Features](https://webappsec.dev/assets/pub/Google_IO-Securing_Web_Apps_with_Modern_Platform_Features.pdf) (Slides)
```

File: src/site/content/en/blog/fetch-metadata/index.md [hunk 10]
```diff
@@ -59,61 +58,74 @@ Requests originating from sites served by your own server (same-origin) will con
...
-A Resource Isolation Policy prevents your resources from being requested by external websites. Blocking such traffic mitigates common cross-site web vulnerabilities such as CSRF, XSSI, timing attacks, and cross-origin information leaks. This policy can be enabled for all endpoints of your application and will allow all resource requests coming from your own application as well as direct navigations (via an HTTP `GET` request). Endpoints that are supposed to be loaded in a cross-site context (e.g. endpoints loaded using CORS) can be opted out of this logic
 
-**To implement a Resource Isolation Policy follow these steps:**
+A Resource Isolation Policy prevents your resources from being requested by external websites. Blocking such traffic mitigates common cross-site web vulnerabilities such as CSRF, XSSI, timing attacks, and cross-origin information leaks. This policy can be enabled for all endpoints of your application and will allow all resource requests coming from your own application as well as direct navigations (via an HTTP `GET` request). Endpoints that are supposed to be loaded in a cross-site context (e.g. endpoints loaded using CORS) can be opted out of this logic.
```
[the remark this answers is not in the document]
Comment (kaycebasques) [reply src/site/content/en/blog/fetch-metadata/index.md:90]: I can't find any grammar articles about "noun phrases being spliced together" and I think the meaning of the sentence is clear enough so let's just leave as is

**Comment by lweichselbaum** [comment]
Thanks everyone! Is there anything else that's missing from my side?

**Pull request merged**